dashboard/lib/devise/models/custom_lockable.rb

```diff
@@ -0,0 +1,44 @@
+module Devise
+  module Models
+    # Extend Devise's built-in Lockable functionality for compatibility with
+    # our particular implementation needs. In particular, we scope it to apply
+    # only to teachers and add SerializedProperties support.
+    #
+    # See https://www.rubydoc.info/github/plataformatec/devise/Devise/Models/Lockable
+    module CustomLockable
```

Comment:
Could we also override the `lock_access!` method to publish an event to Statsig? Product will likely want this metric available for dashboards.

Reply:
Darin and I were just talking about this!

We would indeed like to add more metrics to our general user authentication flow. We were thinking CloudWatch would be a good place, but are we already tracking metrics for some other related things in Statsig?

Reply:
Yeah recently we (platform) have been publishing all of our new backend metrics to Statsig. Product has access to it and generally expects us to put all our metrics there.

It'd be nice to do something like use Cloudwatch as our generic store for logs and metrics, and then pipe them into systems like Statsig/Amplitude (or whatever the next flavor of the month is) using lambdas, but in the meantime it's probably safest to publish backend metrics directly to Statsig. Could also send to both Cloudwatch and Statsig, although that does double up on network calls.

Reply:
I suggest we log these low level system metrics to cloudwatch using our existing MetricsBuffer implementation.

If there is a particular product need to start tracking particular user login events, then we can add in the logging to that system (doubling up), but I would like to avoid storing low level metrics like this in a system that we may need to extract them from down the road.

We have existing CloudWatch metrics in this vein - [User SoftDelete](https://us-east-1.console.aws.amazon.com/cloudwatch/home?region=us-east-1#metricsV2?graph=~(metrics~(~(~'User~'SoftDelete~'Environment~'production~'UserType~'student)~(~'...~'teacher))~view~'timeSeries~stacked~false~region~'us-east-1~start~'-PT1008H~end~'P0D~period~86400~stat~'Sum)&query=~'*7bUser*2cEnvironment*2cUserType*7d). Given that, I would suggest putting these in the `User` namespace, and then metric names like `Locked`, `Unlocked`, `LogIn`, `Create`, etc. The `SoftDelete` metric includes dimensions `environment`, and `UserType`, which I think are appropriate here.

Reply:
@cat5inthecradle that's a great example; didn't realize we had some of these metrics in place already in Cloudwatch.

At the same time, I don't see this as a low-level system metric. If we have a new way to potentially lock out users, I assume product will want to know about it. If we must keep this metric only in Cloudwatch, can we provide read access for platform PM? If we're looking to rework metrics over the next planning year, I suspect PM will need read access anyway, since Cloudwatch makes a ton of sense as the primary store for metrics and logs.

Reply:
I added some overrides to log to both CloudWatch and Statsig, at least for now; we can remove one if we decide to standardize around the other. Please take another look!